### DISCUSSION TITLE
using nix expression to deploy zeek with zeekctl on NixOS or Sandbox

### DISCUSSION THREAD

**Discussion by GTrunSec** [zeek/zeek#939] [How-Tos]
Regarding comments by in https://github.com/zeek/zeek/issues/804#issuecomment-617291437
past few days, I still in a vague conceiving of how to write a nix module of Zeek.

>  I would have suggested changing via configuration options -- there's some you can control through CMake like you found.

First, I can not change `Spool` `Logs` dirs via Cmake options. because of the nix build system doing the process in a sandbox. that why I commented on all of Spool and logs which were in the Cmake list.

Second, the /var/ does not mean my global /var that is `/nix/store/zeek-xx/var` path. That will be made Zeek Cmake can not find `/var` path correctly. 

Third,  that is  `update symlink` process which is related in `nix/store/zeek-xxx/`

my solution and idea:

first, using my [`zeekctl.patch`,](https://github.com/GTrunSec/my-profile/blob/master/own-nixpkgs/zeek/zeekctl.patch) I fixed the dynamic dir problem that pointed Spool and logs in my global `/var/zeek/$PATH`. when I running zeekctl that will be generated in here. (I did it).

> but you can also change what's in zeekctl.cfg to alter various directories. All the options are listed here: https://github.com/zeek/zeekctl#option-reference

second, that was an update symlink problem which you like saw in my before comment.
I'm trying to replace all of zeekctl options on your solution. 

`I will upload test result later`

**Comment by GTrunSec** [comment]
@jsiwek thank you for reviewing my issue.
 what is `working directory` for zeek ? where is or which option?
when I'm using the deploy command with zeekctl, which directory going to create? 

- Fixed zeekctl issue related to https://github.com/zeek/zeek/issues/804#issuecomment-616982060. 
- New Bug output `Error: cannot create working directory for zeek`

* My process and achievement. -  test nix expression of https://github.com/GTrunSec/my-profile/blob/10d18a2862f3f13c229d547674457fb0f6f6555f/own-nixpkgs/zeek/default.nix#L38

1. replace argument of cfgDir  to `/var/zeek/etc`  in /nix/store/xxxx-zeek/etc/zeekctl.cfg
and echo argument of scriptDir  in /nix/store/xxxx-zeek/etc/zeekctl.cfg
```
  --replace "CfgDir = $out/etc" "CfgDir = ${confdir}/etc"    
  echo "scriptsdir = ${confdir}/scripts" >> $out/etc/zeekctl.cfg 
```
2. set hardlinks to/var/db/zeek/scripts path 
```
sudo ln -s /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/*   /var/db/zeek/scripts 
```

 test command
  - deploy
```
sudo ./result/bin/zeekctl deploy                                   
[sudo] password for gtrun: 
checking configurations ...
installing ...
removing old policies in /var/db/zeek/spool/installed-scripts-do-not-touch/site ...
removing old policies in /var/db/zeek/spool/installed-scripts-do-not-touch/auto ...
creating policy directories ...
installing site policies ...
generating standalone-layout.zeek ...
generating local-networks.zeek ...
generating zeekctl-config.zeek ...
generating zeekctl-config.sh ...
stopping ...
stopping zeek ...
starting ...
starting zeek ...
Error: cannot create working directory for zeek
```
  -  diag
```
 sudo ./result/bin/zeekctl diag  
[zeek]
error running crash-diag for zeek
Communication failure with host 127.0.0.1 when checking connection
```
-zeekctl-config.sh
<details>
bindir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/bin"
capstatspath="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/bin/capstats"
cfgdir="/var/db/zeek/etc"
commandtimeout="60"
commtimeout="10"
compresscmd="gzip -9"
compressextension="gz"
compresslogs="1"
controltopic="zeek/control"
crashexpireinterval="0"
croncmd=""
debug="0"
debuglog="/var/db/zeek/spool/debug.log"
defaultstoredir="/var/db/zeek/spool/stores"
env_vars=""
havenfs="0"
helperdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/helpers"
keeplogs=""
lb_custom_interfaceprefix=""
lb_custom_interfacesuffix=""
libdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/lib"
libdir64="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/lib64"
libdirinternal="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/lib/zeekctl"
localnetscfg="/var/db/zeek/etc/networks.cfg"
lockfile="/var/db/zeek/spool/lock"
logdir="/var/db/zeek/logs"
logexpireinterval="0"
logexpireminutes="0"
logrotationinterval="3600"
mailalarmsinterval="86400"
mailalarmsto="root@localhost"
mailarchivelogfail="1"
mailconnectionsummary="1"
mailfrom="Zeek <zeek@nixos>"
mailhostupdown="1"
mailreceivingpackets="1"
mailreplyto=""
mailsubjectprefix="[Zeek]"
mailto="root@localhost"
makearchivename="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/make-archive-name"
memlimit="unlimited"
mindiskspace="5"
nodecfg="/var/db/zeek/etc/node.cfg"
os="Linux"
pcapbufsize="128"
pcapsnaplen="9216"
pfringclusterid="21"
pfringclustertype="4-tuple"
pfringfirstappinstance="0"
pin_command="taskset -c"
plugindir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/lib/zeekctl/plugins"
pluginzeekdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/lib/zeek/plugins"
policydir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeek"
policydirsiteinstall="/var/db/zeek/spool/installed-scripts-do-not-touch/site"
policydirsiteinstallauto="/var/db/zeek/spool/installed-scripts-do-not-touch/auto"
postprocdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/postprocessors"
prefixes="local"
savetraces="0"
scriptsdir="/var/db/zeek/scripts"
sendmail="/nix/store/8cn7xjq8c2jm47nl0xf8d8v3l2bpa2nn-system-sendmail-1.0/bin/sendmail"
sitepluginpath=""
sitepolicypath="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeek/site"
sitepolicyscripts="local.zeek"
spooldir="/var/db/zeek/spool"
standalone="1"
statefile="/var/db/zeek/spool/state.db"
staticdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl"
statsdir="/var/db/zeek/logs/stats"
statslog="/var/db/zeek/spool/stats.log"
statslogenable="1"
statslogexpireinterval="0"
statuscmdshowall="0"
stoptimeout="60"
stopwait="0"
test_enabled="0"
test_foo="1"
time="/run/current-system/sw/bin/time"
timefmt="%d %b %H:%M:%S"
timemachinehost=""
timemachineport="47757/tcp"
tmpdir="/var/db/zeek/spool/tmp"
tmpexecdir="/var/db/zeek/spool/tmp"
tracesummary="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/bin/trace-summary"
version="2.0.0"
zeek="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/bin/zeek"
zeekargs=""
zeekbase="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3"
zeekctlconfigdir="/var/db/zeek/spool"
zeekport="47760"
zeekscriptdir="/nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeek"
</details>
- zeekctl.cfg

```
Global ZeekControl configuration file.

###############################################
# Mail Options

# Recipient address for all emails sent out by Zeek and ZeekControl.
MailTo = root@localhost

# Mail connection summary reports each log rotation interval.  A value of 1
# means mail connection summaries, and a value of 0 means do not mail
# connection summaries.  This option has no effect if the trace-summary
# script is not available.
MailConnectionSummary = 1

# Lower threshold (in percenta``ge of disk space) for space available on the
# disk that holds SpoolDir. If less space is available, "zeekctl cron" starts
# sending out warning emails.  A value of 0 disables this feature.
MinDiskSpace = 5

# Send mail when "zeekctl cron" notices the availability of a host in the
# cluster to have changed.  A value of 1 means send mail when a host status
# changes, and a value of 0 means do not send mail.
MailHostUpDown = 1

###############################################
# Logging Options

# Rotation interval in seconds for log files on manager (or standalone) node.
# A value of 0 disables log rotation.
LogRotationInterval = 3600

# Expiration interval for archived log files in LogDir.  Files older than this
# will be deleted by "zeekctl cron".  The interval is an integer followed by
# one of these time units:  day, hr, min.  A value of 0 means that logs
# never expire.
LogExpireInterval = 0

# Enable ZeekControl to write statistics to the stats.log file.  A value of 1
# means write to stats.log, and a value of 0 means do not write to stats.log.
StatsLogEnable = 1

# Number of days that entries in the stats.log file are kept.  Entries older
# than this many days will be removed by "zeekctl cron".  A value of 0 means
# that entries never expire.
StatsLogExpireInterval = 0

###############################################
# Other Options

# Show all output of the zeekctl status command.  If set to 1, then all output
# is shown.  If set to 0, then zeekctl status will not collect or show the peer
# information (and the command will run faster).
StatusCmdShowAll = 0

# Number of days that crash directories are kept.  Crash directories older
# than this many days will be removed by "zeekctl cron".  A value of 0 means
# that crash directories never expire.
CrashExpireInterval = 0

# Site-specific policy script to load. Zeek will look for this in
# $PREFIX/share/zeek/site. A default local.zeek comes preinstalled
# and can be customized as desired.
SitePolicyScripts = local.zeek

# Location of the log directory where log files will be archived each rotation
# interval.
LogDir = /var/db/zeek/logs

# Location of the spool directory where files and data that are currently being
# written are stored.
SpoolDir = /var/db/zeek/spool

# Location of other configuration files that can be used to customize
# ZeekControl operation (e.g. local networks, nodes).
CfgDir = /var/db/zeek/etc

scriptsdir = /var/db/zeek/scripts
```
-  /var/db/zeek tree structure


```
.
├── etc
│   ├── networks.cfg
│   └── node.cfg
├── logs
│   └── current -> /var/db/zeek/spool/zeek
├── scripts
├── scripts                                                                                                                              
│   ├── archive-log -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/archive-log                          
│   ├── check-config ->  /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/check-config                                                                                                                 
│   ├── crash-diag -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/crash-diag                            
│   ├── delete-log -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/delete-log                            
│   ├── expire-crash -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/expire-crash                        
│   ├── expire-logs -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/expire-logs                          
│   ├── helpers -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/helpers                                  
│   ├── make-archive-name -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/make-archive-name              
│   ├── postprocessors -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/postprocessors                    
│   ├── post-terminate -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/post-terminate                    
│   ├── run-zeek -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/run-zeek
│   ├── run-zeek-on-trace -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/run-zeek-on-trace
│   ├── send-mail -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/send-mail
│   ├── set-zeek-path -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/set-zeek-path
│   ├── stats-to-csv -> /nix/store/pcj8z67f2a4w26azgvc3kq6fhz7i6a9i-zeek-3.0.3/share/zeekctl/scripts/stats-to-csv
│   └── zeekctl-config.sh -> /var/db/zeek/spool/zeekctl-config.sh
├── spool
│   ├── installed-scripts-do-not-touch
│   │   ├── auto
│   │   │   ├── local-networks.zeek
│   │   │   ├── standalone-layout.zeek
│   │   │   └── zeekctl-config.zeek
│   │   └── site
│   │       └── local.zeek
│   ├── state.db
│   ├── tmp
│   └── zeekctl-config.sh
├── state.db
└── tmp
```

**Comment by GTrunSec** [comment]
I think that it should be a good approach to install Zeek's plugins with a JSON list. https://github.com/hardenedlinux/debian-nix-manager/blob/c7ed0e94147f2b01d46f85fc9cc1e83fa298be53/pkgs/zeek/default.nix#L6
```
00:48:09 with gtrun in nixpkgs/own-nixpkgs/zeek at nixos on  master [$✘] 
➜ sudo ./result/bin/zeek  -N Apache::Kafka && sudo ./result/bin/zeek  -N Johanna::PostgreSQL
Apache::Kafka - Writes logs to Kafka (dynamic, version 0.3.0)
Johanna::PostgreSQL - PostgreSQL log writer and input reader (dynamic, version 0.2.0)
```
In addition, I'm going to write a Nix module for Zeek, such as create directories and set hard links automatically and setting arguments with nix expression in Zeek cfg files(node.cfg, network.cfg, etc.).

Surely, if the deploy problem can be solved in the next few days.

**Comment by jsiwek** [comment]
> what is `working directory` for zeek ? where is or which option?

I don't know, you may find out more quickly by extending that error output to include the failed path: https://github.com/zeek/zeekctl/blob/7e65a34905ec9684c442da5f737fe75beb94aae6/ZeekControl/control.py#L175

**Comment by GTrunSec** [comment]
> github.com/zeek/zeekctl/blob/7e65a34905ec9684c442da5f737fe75beb94aae6/ZeekControl/control.py#L175

Thanks!
I have solved it. 

- added new cfg to zeekctl.cfg
`echo "helperdir = ${confdir}/scripts/helpers" >> $out/etc/zeekctl.cfg `

- Do not set hard links for these dirs.

```
sudo cp -r /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeekctl/scripts/helpers /var/db/zeek/scripts/

sudo cp -r /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeekctl/scripts/postprocessors /var/db/zeek/scripts/
```

- then set each of hard link to  /var/db/ (without /helpers and /postprocessors)

```
sudo ln -s /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeekctl/scripts/* /var/db/zeek/scripts/

```

Check the status:  it's working.
```
sudo zeekctl status                                                                                            
Name         Type       Host          Status    Pid    Started
zeek         standalone localhost     running   30561  30 Apr 15:29:12
```
- cat /var/db/zeek/logs/current/loaded_scripts.log:  until now, it looks like good.
```
  /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/policy/frameworks/notice/extend-email/hostnames.zeek
/nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/zeekctl/__load__.zeek
  /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/zeekctl/main.zeek
    /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/policy/frameworks/control/controllee.zeek
/nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/zeekctl/standalone.zeek
  /var/db/zeek/spool/installed-scripts-do-not-touch/auto/standalone-layout.zeek
  /nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/policy/misc/trim-trace-file.zeek
/nix/store/m44d96wwy0bizmqdvmivljf0kffffhqf-zeek-3.0.5/share/zeek/zeekctl/auto.zeek
  /var/db/zeek/spool/installed-scripts-do-not-touch/auto/local-networks.zeek
  /var/db/zeek/spool/installed-scripts-do-not-touch/auto/zeekctl-config.zeek
```

**Comment by GTrunSec** [comment]
this issue can be useful and traceable who would like running or deploying Zeek in the sandbox.

**Comment by GTrunSec** [comment]
this solution can not deploy on NixOS. https://github.com/NixOS/nixpkgs/pull/86494#issuecomment-622777999 . but working well on Debian (OS) https://github.com/hardenedlinux/debian-nix-manager/blob/master/modules/zeek.nix
tracking error output : `Error: cannot create working directory for zeek`.

**Comment by GTrunSec** [comment]
https://github.com/NixOS/nixpkgs/pull/86494#issuecomment-623024864

`Error: cannot create working directory for Zeek.`

the Spool/zeek couldn't be created when running deploy command with Zeekctl.
it will be generated an invalid hard link of `Spool/zeek` to `logs/current` .
